File: util/pack/NyCastle.java
package common.util.pack;

import common.CommonStatic;
import common.CommonStatic.BCAuxAssets;
import common.system.VImg;
import common.system.fake.FakeImage;
import common.util.anim.*;

public class NyCastle extends AnimI<NyCastle, NyCastle.NyType> {

	public enum NyType implements AnimI.AnimType<NyCastle, NyType> {
		BASE, ATK, EXT
	}

	public static final int TOT = 8;

	public static void read() {
		BCAuxAssets aux = CommonStatic.getBCAssets();
		String pre = "./org/castle/00";
		String mid = "/nyankoCastle_00";
		int[] type = new int[] { 0, 2, 3 };
		for (int t = 0; t < 3; t++)
			for (int i = 0; i < TOT; i++) {
				String str = pre + type[t] + mid + type[t] + "_0" + i;
				aux.main[t][i] = new VImg(str + ".png");
			}
		for (int i = 0; i < TOT; i++) {
			String str = pre + 1 + mid + 1 + "_0";
			aux.atks[i] = new NyCastle(str, i);
		}
	}

	private final int id;
	private final VImg sprite;
	private final ImgCut ic;
	private final MaModel model, atkm, extm;
	private final MaAnim manim, atka, exta;
	private FakeImage[] parts;

	private NyCastle(String str, int t) {
		anim = this;
		id = t;
		sprite = new VImg(str + t + "_00.png");
		ic = ImgCut.newIns(str + t + "_00.imgcut");
		model = MaModel.newIns(str + t + "_01.mamodel");
		manim = MaAnim.newIns(str + t + "_01.maanim");
		if (t != 1 && t != 2 && t != 7) {
			atkm = MaModel.newIns(str + t + "_00.mamodel");
			atka = MaAnim.newIns(str + t + "_00.maanim");
		} else {
			atkm = null;
			atka = null;
		}
		if (t == 6) {
			extm = MaModel.newIns(str + t + "_02.mamodel");
			exta = MaAnim.newIns(str + t + "_02.maanim");
		} else {
			extm = null;
			exta = null;
		}
	}

	@Override
	public void check() {
		if (parts == null)
			load();
	}

	@Override
	public EAnimD<NyType> getEAnim(NyType t) {
		check();
		if (t == NyType.BASE)
			return new EAnimD<>(this, model, manim, t);
		if (t == NyType.ATK)
			return new EAnimD<>(this, atkm, atka, t);
		if (t == NyType.EXT)
			return new EAnimD<>(this, extm, exta, t);
		return null;
	}

	@Override
	public void load() {
		parts = ic.cut(sprite.getImg());
	}

	@Override
	public String[] names() {
		if (atkm == null)
			return new String[] { "castle" };
		if (extm == null)
			return new String[] { "castle", "atk" };
		return new String[] { "castle", "atk", "ext" };
	}

	@Override
	public FakeImage parts(int i) {
		return parts[i];
	}

	@Override
	public String toString() {
		return "castle " + id;
	}

	@Override
	public NyType[] types() {
		return NyType.values();
	}

}

File: util/pack/NyCastle.kt
package common.util.pack

import common.CommonStatic
import common.CommonStatic.BCAuxAssets
import common.system.VImg
import common.system.fake.FakeImage
import common.util.anim.*
import common.util.pack.NyCastle
import common.util.pack.NyCastle.NyType

class NyCastle private constructor(str: String, t: Int) : AnimI<NyCastle?, NyType?>() {
    enum class NyType : AnimType<NyCastle?, NyType?> {
        BASE, ATK, EXT
    }

    private val id: Int
    private val sprite: VImg
    private val ic: ImgCut
    private val model: MaModel
    private var atkm: MaModel? = null
    private var extm: MaModel? = null
    private val manim: MaAnim
    private var atka: MaAnim? = null
    private var exta: MaAnim? = null
    private var parts: Array<FakeImage>?
    override fun check() {
        if (parts == null) load()
    }

    override fun getEAnim(t: NyType): EAnimD<NyType>? {
        check()
        if (t == NyType.BASE) return EAnimD<NyType>(this, model, manim)
        if (t == NyType.ATK) return EAnimD<NyType>(this, atkm, atka)
        return if (t == NyType.EXT) EAnimD<NyType>(this, extm, exta) else null
    }

    override fun load() {
        parts = ic.cut(sprite.getImg())
    }

    override fun names(): Array<String?> {
        if (atkm == null) return arrayOf("castle")
        return if (extm == null) arrayOf("castle", "atk") else arrayOf("castle", "atk", "ext")
    }

    override fun parts(i: Int): FakeImage? {
        return parts!![i]
    }

    override fun toString(): String {
        return "castle $id"
    }

    override fun types(): Array<NyType> {
        return common.util.pack.NyCastle.NyType.values()
    }

    companion object {
        const val TOT = 8
        fun read() {
            val aux: BCAuxAssets = CommonStatic.getBCAssets()
            val pre = "./org/castle/00"
            val mid = "/nyankoCastle_00"
            val type = intArrayOf(0, 2, 3)
            for (t in 0..2) for (i in 0 until TOT) {
                val str = pre + type[t] + mid + type[t] + "_0" + i
                aux.main.get(t).get(i) = VImg("$str.png")
            }
            for (i in 0 until TOT) {
                val str = pre + 1 + mid + 1 + "_0"
                aux.atks.get(i) = NyCastle(str, i)
            }
        }
    }

    init {
        anim = this
        id = t
        sprite = VImg(str + t + "_00.png")
        ic = ImgCut.Companion.newIns(str + t + "_00.imgcut")
        model = MaModel.Companion.newIns(str + t + "_01.mamodel")
        manim = MaAnim.Companion.newIns(str + t + "_01.maanim")
        if (t != 1 && t != 2 && t != 7) {
            atkm = MaModel.Companion.newIns(str + t + "_00.mamodel")
            atka = MaAnim.Companion.newIns(str + t + "_00.maanim")
        } else {
            atkm = null
            atka = null
        }
        if (t == 6) {
            extm = MaModel.Companion.newIns(str + t + "_02.mamodel")
            exta = MaAnim.Companion.newIns(str + t + "_02.maanim")
        } else {
            extm = null
            exta = null
        }
    }
}
